File: bridge/swiftui/Sources/NUISwiftUIBridge/Primitive.swift
indirect enum Primitive: Codable, Hashable {
    case empty
    case text(content: String)
    case button(label: Id<Primitive>)
    case vStack(wrapped: Id<Primitive>)
    case hStack(wrapped: Id<Primitive>)
    case zStack(wrapped: Id<Primitive>)
    case tuple2(child1: Id<Primitive>, child2: Id<Primitive>)
    case tuple3(child1: Id<Primitive>, child2: Id<Primitive>, child3: Id<Primitive>)
    case tuple4(child1: Id<Primitive>, child2: Id<Primitive>, child3: Id<Primitive>, child4: Id<Primitive>)
    case tuple5(child1: Id<Primitive>, child2: Id<Primitive>, child3: Id<Primitive>, child4: Id<Primitive>, child5: Id<Primitive>)
}

File: shared/src/primitive.rs
use serde::{Serialize, Deserialize};

/// A UI component tree composed of primitive components.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Primitive {
    Empty,
    Button { label: Box<Primitive> },
    Text { content: String },
    VStack { wrapped: Box<Primitive> },
    HStack { wrapped: Box<Primitive> },
    ZStack { wrapped: Box<Primitive> },
    Tuple2 { child1: Box<Primitive>, child2: Box<Primitive> },
    Tuple3 { child1: Box<Primitive>, child2: Box<Primitive>, child3: Box<Primitive> },
    Tuple4 { child1: Box<Primitive>, child2: Box<Primitive>, child3: Box<Primitive>, child4: Box<Primitive> },
    Tuple5 { child1: Box<Primitive>, child2: Box<Primitive>, child3: Box<Primitive>, child4: Box<Primitive>, child5: Box<Primitive> },
}
